File: Day-20/cqrs-sample/python/cqrs_order_system.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum


class OrderStatus(Enum):
    CREATED = "CREATED"
    CONFIRMED = "CONFIRMED"


@dataclass(frozen=True)
class OrderAggregate:
    order_id: str
    customer_id: str
    amount_cents: int
    status: OrderStatus


@dataclass(frozen=True)
class DomainEvent:
    sequence: int
    event_type: str
    order_id: str
    customer_id: str
    amount_cents: int


@dataclass
class OrderReadModel:
    order_id: str
    customer_id: str
    amount_cents: int
    status: str


class WriteStore:
    def __init__(self) -> None:
        self.orders: dict[str, OrderAggregate] = {}
        self.events: list[DomainEvent] = []
        self._next_sequence = 1

    def append_event(
        self, event_type: str, order_id: str, customer_id: str, amount_cents: int
    ) -> DomainEvent:
        event = DomainEvent(
            sequence=self._next_sequence,
            event_type=event_type,
            order_id=order_id,
            customer_id=customer_id,
            amount_cents=amount_cents,
        )
        self._next_sequence += 1
        self.events.append(event)
        return event


class ReadStore:
    def __init__(self) -> None:
        self.orders: dict[str, OrderReadModel] = {}


class CommandService:
    def __init__(self, write_store: WriteStore) -> None:
        self._write_store = write_store

    def create_order(self, order_id: str, customer_id: str, amount_cents: int) -> None:
        if order_id in self._write_store.orders:
            raise ValueError(f"order already exists: {order_id}")
        if amount_cents <= 0:
            raise ValueError("amount_cents must be positive")

        aggregate = OrderAggregate(
            order_id=order_id,
            customer_id=customer_id,
            amount_cents=amount_cents,
            status=OrderStatus.CREATED,
        )
        self._write_store.orders[order_id] = aggregate
        self._write_store.append_event(
            event_type="OrderCreated",
            order_id=order_id,
            customer_id=customer_id,
            amount_cents=amount_cents,
        )

    def confirm_order(self, order_id: str) -> None:
        aggregate = self._write_store.orders.get(order_id)
        if aggregate is None:
            raise ValueError(f"order not found: {order_id}")
        if aggregate.status == OrderStatus.CONFIRMED:
            return

        confirmed = OrderAggregate(
            order_id=aggregate.order_id,
            customer_id=aggregate.customer_id,
            amount_cents=aggregate.amount_cents,
            status=OrderStatus.CONFIRMED,
        )
        self._write_store.orders[order_id] = confirmed
        self._write_store.append_event(
            event_type="OrderConfirmed",
            order_id=order_id,
            customer_id=aggregate.customer_id,
            amount_cents=aggregate.amount_cents,
        )


class Projector:
    def __init__(self, write_store: WriteStore, read_store: ReadStore) -> None:
        self._write_store = write_store
        self._read_store = read_store
        self._last_sequence = 0

    @property
    def last_sequence(self) -> int:
        return self._last_sequence

    def project_next(self, batch_size: int = 10) -> list[str]:
        history: list[str] = []
        if batch_size <= 0:
            return history

        unapplied = [e for e in self._write_store.events if e.sequence > self._last_sequence]
        for event in unapplied[:batch_size]:
            if event.event_type == "OrderCreated":
                self._read_store.orders[event.order_id] = OrderReadModel(
                    order_id=event.order_id,
                    customer_id=event.customer_id,
                    amount_cents=event.amount_cents,
                    status=OrderStatus.CREATED.value,
                )
                history.append(f"projected {event.event_type}#{event.sequence}")
            elif event.event_type == "OrderConfirmed":
                row = self._read_store.orders.get(event.order_id)
                if row is not None:
                    row.status = OrderStatus.CONFIRMED.value
                history.append(f"projected {event.event_type}#{event.sequence}")
            else:
                history.append(f"skipped unknown event#{event.sequence}")
            self._last_sequence = event.sequence

        return history


class QueryService:
    def __init__(self, read_store: ReadStore) -> None:
        self._read_store = read_store

    def get_order(self, order_id: str) -> OrderReadModel | None:
        row = self._read_store.orders.get(order_id)
        if row is None:
            return None
        return OrderReadModel(
            order_id=row.order_id,
            customer_id=row.customer_id,
            amount_cents=row.amount_cents,
            status=row.status,
        )

    def list_orders_for_customer(self, customer_id: str) -> list[OrderReadModel]:
        rows = [r for r in self._read_store.orders.values() if r.customer_id == customer_id]
        return [
            OrderReadModel(
                order_id=r.order_id,
                customer_id=r.customer_id,
                amount_cents=r.amount_cents,
                status=r.status,
            )
            for r in rows
        ]

File: Day-20/cqrs-sample/java/CqrsOrderSystem.java
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class CqrsOrderSystem {
    public enum OrderStatus {
        CREATED,
        CONFIRMED
    }

    public static final class OrderAggregate {
        private final String orderId;
        private final String customerId;
        private final int amountCents;
        private final OrderStatus status;

        public OrderAggregate(String orderId, String customerId, int amountCents, OrderStatus status) {
            this.orderId = orderId;
            this.customerId = customerId;
            this.amountCents = amountCents;
            this.status = status;
        }

        public String getOrderId() {
            return orderId;
        }

        public String getCustomerId() {
            return customerId;
        }

        public int getAmountCents() {
            return amountCents;
        }

        public OrderStatus getStatus() {
            return status;
        }
    }

    public static final class DomainEvent {
        private final int sequence;
        private final String eventType;
        private final String orderId;
        private final String customerId;
        private final int amountCents;

        public DomainEvent(
            int sequence,
            String eventType,
            String orderId,
            String customerId,
            int amountCents
        ) {
            this.sequence = sequence;
            this.eventType = eventType;
            this.orderId = orderId;
            this.customerId = customerId;
            this.amountCents = amountCents;
        }

        public int getSequence() {
            return sequence;
        }

        public String getEventType() {
            return eventType;
        }

        public String getOrderId() {
            return orderId;
        }

        public String getCustomerId() {
            return customerId;
        }

        public int getAmountCents() {
            return amountCents;
        }
    }

    public static final class OrderReadModel {
        private final String orderId;
        private final String customerId;
        private final int amountCents;
        private String status;

        public OrderReadModel(String orderId, String customerId, int amountCents, String status) {
            this.orderId = orderId;
            this.customerId = customerId;
            this.amountCents = amountCents;
            this.status = status;
        }

        public String getOrderId() {
            return orderId;
        }

        public String getCustomerId() {
            return customerId;
        }

        public int getAmountCents() {
            return amountCents;
        }

        public String getStatus() {
            return status;
        }

        public void setStatus(String status) {
            this.status = status;
        }
    }

    public static final class WriteStore {
        private final Map<String, OrderAggregate> orders = new LinkedHashMap<>();
        private final List<DomainEvent> events = new ArrayList<>();
        private int nextSequence = 1;

        public DomainEvent appendEvent(
            String eventType,
            String orderId,
            String customerId,
            int amountCents
        ) {
            DomainEvent event = new DomainEvent(
                nextSequence,
                eventType,
                orderId,
                customerId,
                amountCents
            );
            nextSequence += 1;
            events.add(event);
            return event;
        }

        public Map<String, OrderAggregate> getOrders() {
            return orders;
        }

        public List<DomainEvent> getEvents() {
            return events;
        }
    }

    public static final class ReadStore {
        private final Map<String, OrderReadModel> orders = new LinkedHashMap<>();

        public Map<String, OrderReadModel> getOrders() {
            return orders;
        }
    }

    public static final class CommandService {
        private final WriteStore writeStore;

        public CommandService(WriteStore writeStore) {
            this.writeStore = writeStore;
        }

        public void createOrder(String orderId, String customerId, int amountCents) {
            if (writeStore.getOrders().containsKey(orderId)) {
                throw new IllegalArgumentException("order already exists: " + orderId);
            }
            if (amountCents <= 0) {
                throw new IllegalArgumentException("amountCents must be positive");
            }

            OrderAggregate aggregate = new OrderAggregate(
                orderId,
                customerId,
                amountCents,
                OrderStatus.CREATED
            );
            writeStore.getOrders().put(orderId, aggregate);
            writeStore.appendEvent("OrderCreated", orderId, customerId, amountCents);
        }

        public void confirmOrder(String orderId) {
            OrderAggregate aggregate = writeStore.getOrders().get(orderId);
            if (aggregate == null) {
                throw new IllegalArgumentException("order not found: " + orderId);
            }
            if (aggregate.getStatus() == OrderStatus.CONFIRMED) {
                return;
            }

            OrderAggregate confirmed = new OrderAggregate(
                aggregate.getOrderId(),
                aggregate.getCustomerId(),
                aggregate.getAmountCents(),
                OrderStatus.CONFIRMED
            );
            writeStore.getOrders().put(orderId, confirmed);
            writeStore.appendEvent(
                "OrderConfirmed",
                orderId,
                aggregate.getCustomerId(),
                aggregate.getAmountCents()
            );
        }
    }

    public static final class Projector {
        private final WriteStore writeStore;
        private final ReadStore readStore;
        private int lastSequence = 0;

        public Projector(WriteStore writeStore, ReadStore readStore) {
            this.writeStore = writeStore;
            this.readStore = readStore;
        }

        public int getLastSequence() {
            return lastSequence;
        }

        public List<String> projectNext(int batchSize) {
            List<String> history = new ArrayList<>();
            if (batchSize <= 0) {
                return history;
            }

            int processed = 0;
            for (DomainEvent event : writeStore.getEvents()) {
                if (event.getSequence() <= lastSequence) {
                    continue;
                }
                if (processed >= batchSize) {
                    break;
                }

                if ("OrderCreated".equals(event.getEventType())) {
                    readStore.getOrders().put(
                        event.getOrderId(),
                        new OrderReadModel(
                            event.getOrderId(),
                            event.getCustomerId(),
                            event.getAmountCents(),
                            OrderStatus.CREATED.name()
                        )
                    );
                    history.add("projected " + event.getEventType() + "#" + event.getSequence());
                } else if ("OrderConfirmed".equals(event.getEventType())) {
                    OrderReadModel row = readStore.getOrders().get(event.getOrderId());
                    if (row != null) {
                        row.setStatus(OrderStatus.CONFIRMED.name());
                    }
                    history.add("projected " + event.getEventType() + "#" + event.getSequence());
                } else {
                    history.add("skipped unknown event#" + event.getSequence());
                }

                lastSequence = event.getSequence();
                processed += 1;
            }

            return history;
        }
    }

    public static final class QueryService {
        private final ReadStore readStore;

        public QueryService(ReadStore readStore) {
            this.readStore = readStore;
        }

        public OrderReadModel getOrder(String orderId) {
            OrderReadModel row = readStore.getOrders().get(orderId);
            if (row == null) {
                return null;
            }
            return new OrderReadModel(
                row.getOrderId(),
                row.getCustomerId(),
                row.getAmountCents(),
                row.getStatus()
            );
        }

        public List<OrderReadModel> listOrdersForCustomer(String customerId) {
            List<OrderReadModel> result = new ArrayList<>();
            for (OrderReadModel row : readStore.getOrders().values()) {
                if (!customerId.equals(row.getCustomerId())) {
                    continue;
                }
                result.add(
                    new OrderReadModel(
                        row.getOrderId(),
                        row.getCustomerId(),
                        row.getAmountCents(),
                        row.getStatus()
                    )
                );
            }
            return result;
        }
    }
}
